### ISSUE TITLE
Argo-workflows missing permissions in ClusterRole *-cluster-template

### ISSUE THREAD

**Issue by xmsanchez** [argoproj/argo-helm#1624]
### Describe the bug

When the argo-workflows-server pods start it states that there are missing permissions:

```
E1107 16:00:29.872140       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.24.3/tools/cache/reflector.go:167: Failed to watch *v1.ServiceAccount: failed to list *v1.ServiceAccount: serviceaccounts is forbidden: User "system:serviceaccount:argo-workflows:argo-workflows-server" cannot list resource "serviceaccounts" in API group "" at the cluster scope
```

The issue is in this role: https://github.com/argoproj/argo-helm/blob/48360eeb953fcacf5c6298df2ce35c05d1052858/charts/argo-workflows/templates/server/server-cluster-roles.yaml#L123

The following permissions fixes the above error:

```
- apiGroups:
  - ""
  resources:
  - serviceaccounts
  verbs:
  - get
  - list
  - watch
```

Just patching the ClusterRole with above permissions do work.

Found this issue in the version `0.20.6` of the helm chart. I believe older versions were working properly, but I'm not sure.

### Related helm chart

argo-workflows

### Helm chart version

0.20.6

### To Reproduce

Deploy argo-workflows helm chart with `server.rbac.create: true`.

Inspect the argo-workflows-server pod. Never gets up and logs show the errors.

### Expected behavior

The pod should get up & running right away.

**Comment by yu-croco** [comment]
Hi @xmsanchez , thank you for reporting.
I checked `v0.20.6` on my local environment with `server.rbac.create: true` , but I did not get any errors in `argo-workflows-server` pod. 
Are there any additional things that I have to do for reproducing the issue? 🤔 

```
$ k get po argo-argo-workflows-server-c67fc666c-ddrdv
NAME                                         READY   STATUS    RESTARTS   AGE
argo-argo-workflows-server-c67fc666c-ddrdv   1/1     Running   0          57s
```

```
$ k logs argo-argo-workflows-server-c67fc666c-ddrdv -f
time="2022-11-11T04:02:43.713Z" level=info msg="not enabling pprof debug endpoints"
time="2022-11-11T04:02:43.714Z" level=info authModes="[client]" baseHRef=/ managedNamespace= namespace=default secure=false ssoNamespace=default
time="2022-11-11T04:02:43.714Z" level=warning msg="You are running in insecure mode. Learn how to enable transport layer security: https://argoproj.github.io/argo-workflows/tls/"
time="2022-11-11T04:02:43.714Z" level=info msg="SSO disabled"
time="2022-11-11T04:02:43.717Z" level=info msg="Starting Argo Server" instanceID= version=v3.4.3
time="2022-11-11T04:02:43.717Z" level=info msg="Creating event controller" asyncDispatch=false operationQueueSize=16 workerCount=4
time="2022-11-11T04:02:43.719Z" level=info msg="GRPC Server Max Message Size, MaxGRPCMessageSize, is set" GRPC_MESSAGE_SIZE=104857600
time="2022-11-11T04:02:43.719Z" level=info msg="Argo Server started successfully on http://localhost:2746" url="http://localhost:2746"
time="2022-11-11T04:03:05.721Z" level=info duration="487.042µs" method=GET path=index.html size=473 status=0
time="2022-11-11T04:03:25.721Z" level=info duration="319.958µs" method=GET path=index.html size=473 status=0
```

**Comment by xmsanchez** [comment]
Hi @yu-croco , I've tailored the issue a little bit more and the issue happens when using auth-mode=sso. 

My values file was extensive, I apologize for not being more specific. Here's how to reproduce it:

```
server:
  enabled: true
  extraArgs:
  - --auth-mode=sso
  secure: true
  sso:
    clientId:
      key: client-id
      name: argo-workflows-oauth
    clientSecret:
      key: client-secret
      name: argo-workflows-oauth
    issuer: https://accounts.google.com
    redirectUrl: https://argowf.example.com/oauth2/callback
```

**Comment by yu-croco** [comment]
@xmsanchez 
Thank you for the detail, I reproduced it. 😄

**Comment by yu-croco** [comment]
@xmsanchez 
I investigated it. I think you need to add [server.sso.rbac.enabled](https://github.com/argoproj/argo-helm/blob/main/charts/argo-workflows/values.yaml#L502-L503) block, like below.

```yaml
server:
  enabled: true
  extraArgs:
  - --auth-mode=sso
  secure: true
  sso:
    clientId:
      key: client-id
      name: argo-workflows-oauth
    clientSecret:
      key: client-secret
      name: argo-workflows-oauth
    issuer: https://accounts.google.com
    redirectUrl: https://argowf.example.com/oauth2/callback
    rbac: # <- add this block
      enabled: true
```

Once you add it, argo-workflow-server is granted serviceaccount permission, [here](https://github.com/argoproj/argo-helm/blob/main/charts/argo-workflows/templates/server/server-cluster-roles.yaml#L49-L58), so argo-workflow-server would run correctly. 🙋

**Comment by smolinari** [comment]
I'd just like to note. I just upgraded to 0.29.1 from the latest 0.28.x and didn't have the rbac enabled as shown above, and the upgrade failed with this OP's error. I set `rbac` to true and the upgrade worked.

 ```yaml
     rbac: # <- add this block
      enabled: true # <- add this block
 ```
 
 Scott

**Comment by signaleleven** [comment]
I came from a situation where I had `sso` enabled and `rbac` disabled. This is still possible with 0.28.2, but 0.29 ties to having rbac enabled the condition to let `workflow-server` access `ServiceAccounts`.  (see https://github.com/argoproj/argo-helm/blob/b8c29f7db6ba38126551c1789a78763e702cde62/charts/argo-workflows/templates/server/server-cluster-roles.yaml#L57C1-L66)

AFAICT that permission is necessary also when rbac is disabled. Before 0.29 the template created it when rbac was defined (and indeed I had it defined in my values as `enabled: false`"). 

Should a new issue be raised for this? I would make a PR but I'm not entirely sure if having that permission (get serviceaccounts) enabled in every case is good.

**Comment by driv** [comment]
I'm having the same issue, I can't have `sso` enabled without `rbac`.
All of the fixes I've found is to just enable rbac.

I'm not sure if this is an issue with the chart or with argo workflows that it's trying to read the `serviceaccounts` despite `rbac` being disabled.

**Comment by agilgur5** [comment]
> I'm having the same issue, I can't have `sso` enabled without `rbac`.

Thought I'd mention here that this is indeed a bug. You should be able to enable `sso` without `sso.rbac`. #2159 ended up duplicating this issue as such.

 I root caused this in https://github.com/argoproj/argo-helm/issues/2159#issuecomment-1646988596 and turns out that this is actually a bug upstream in Argo itself. Filed a PR upstream to fix this: https://github.com/argoproj/argo-workflows/pull/11426

**Comment by agilgur5** [comment]
Upstream fix will be available in v3.4.10+: https://github.com/argoproj/argo-workflows/issues/11552
